### PR TITLE
Marginalize out discrete variables in the model

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #915 
Pair coded with @eb8680 

This supports exact marginalization out discrete variables that may appear in the model but not the guide. For example to infer parameters of an HMM in Bayesian Baum-Welch inference, we can
```py
def model(data):
    dim = data.shape[-1]
    jeffreys_prior = dist.Dirichlet(0.5 * torch.ones(dim, dim)).independent(1)
    trans = pyro.sample("trans", jeffreys_prior)
    emit = pyro.sample("emit", jeffreys_prior)
    x = 0
    for t, y in enumerate(data):
        x = pyro.sample("x_{}".format(t), dist.Categorical(trans[x]),
                        infer={"enumerate": "parallel", "expand": False})
        pyro.sample("y_{}".format(t), dist.Categorical(emit[x]), obs=y)

guide = AutoMultivariateNormal(poutine.block(model, expose=["trans", "emit"]))
```
A follow-up PR will add this example as a tutorial.

## Design choices

This adds model enumeration logic to `TraceEnum_ELBO`. @martinjankowiak suggested a cleaner alternative design that modifies arbitrary models, with an interface closer to either `poutine` or a `Distribution`. Ideally we can refactor to this more general interface later. For now I'd like to merge this working implementation in `TraceEnum_ELBO` to enable modeling and to serve as a reference implementation for the more general version.

## Tested

- [x] added smoke tests of the HMM in the model
- [x] added loss-and-grads tests for various models, including `iarange`
- [x] added validity and error tests to `test_valid_models.py`

### CODE REVIEW

**Comment by martinjankowiak** [comment]
nice test coverage 💯

File: pyro/infer/traceenum_elbo.py [hunk 1]
```diff
@@ -10,25 +10,65 @@
...
+            logprobs_t = sum((logprobs_u for u, logprobs_u in enum_logprobs.items() if u <= t), [])
+            if not logprobs_t:
+                marginal_costs[t] = costs_t
+                continue
+
+            ordinal_shape = broadcast_shape(*(x.shape for x in logprobs_t))
+            for cost_t in costs_t:
+                target_shape = broadcast_shape(cost_t.shape, ordinal_shape)
+                target_shape = target_shape[enum_boundary:] if enum_boundary else ()
+                print('{}: target_shape={}, factor shapes = {}'.format(
```
Comment (neerajprad) [inline pyro/infer/traceenum_elbo.py]: Can we remove this print statement, or add it as `debug` statement using `pyro.log.debug` instead? Our test logs are increasing in size, and it will be nice to be able to toggle debug statements if needed.

File: pyro/infer/traceenum_elbo.py [hunk 2]
```diff
@@ -83,9 +124,16 @@ def _get_traces(self, model, guide, *args, **kwargs):
...
-        # enable parallel enumeration over the vectorized guide.
-        guide = poutine.enum(guide, first_available_dim=self.max_iarange_nesting)
+        else:
+            guide = poutine.broadcast(guide)
+            model = poutine.broadcast(model)
+
+        # Enable parallel enumeration over the vectorized guide and model.
+        # The model allocates enumeration dimensions after (to the left of) the guide.
+        guide_enum = EnumerateMessenger(first_available_dim=self.max_iarange_nesting)
+        model_enum = EnumerateMessenger(first_available_dim=lambda: guide_enum.next_available_dim)
```
Comment (neerajprad) [inline pyro/infer/traceenum_elbo.py]: Interesting use of deferred evaluation to start the model enum block. :)

File: pyro/infer/traceenum_elbo.py [hunk 3]
```diff
@@ -10,25 +10,65 @@
...
+# TODO move this logic into a poutine
+def _compute_model_costs(model_trace, guide_trace, ordering):
+    # Collect model log_probs, possibly marginalizing out enumerated model variables.
+    costs = OrderedDict()
+    enum_logprobs = OrderedDict()
+    enum_dims = []
+    for name, site in model_trace.nodes.items():
+        if site["type"] == "sample":
+            site["log_prob"]._pyro_name = name  # DEBUG
+            if name in guide_trace or not site["infer"].get("_enumerate_dim") is not None:
```
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/infer/traceenum_elbo.py]: No, because `poutine.replay` copies `site["infer"]` from the guide. (We tried your idea initially :smile:)

**Comment by fritzo** [comment]
I believe this is ready to merge. I'll update #1329 as soon as this merges.
